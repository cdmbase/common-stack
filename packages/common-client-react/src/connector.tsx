import * as React from 'react';

import { AbstractFeature, IFeature } from '@common-stack/client-core';


export class Feature extends AbstractFeature implements IFeature {
    get routes() {
        return this.route.map((component, idx) => React.cloneElement(component, { key: idx + this.route.length }));
    }

    get navItems() {
        return this.navItem.map((component, idx) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    get navItemsRight() {
        return this.navItemRight.map((component, idx) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    public getWrappedRoot(root, req) {
        let nestedRoot = root;
        for (const componentFactory of this.rootComponentFactory) {
            nestedRoot = React.cloneElement(componentFactory(req), {}, nestedRoot);
        }
        return nestedRoot;
    }

}
