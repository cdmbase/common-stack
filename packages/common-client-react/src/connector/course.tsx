import * as React from 'react';
import { Link } from 'react-router-dom';

interface MyProps {
  history?: any;
  children?: any;
  key?: any;
};

const Course: React.StatelessComponent<MyProps> = (props: MyProps) => {
  const from = "localhost:3000/"; 
  return (
  <div key={props.history.id}>
    <Link to={`${props.history.number}`} {...from}>{`Page: ${
        props.history.number
      }`}</Link>
  </div>
)};

export default Course;