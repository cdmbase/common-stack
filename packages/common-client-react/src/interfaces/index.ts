export * from './router';
export * from './extended-feature';
export * from './menu';
export * from './new-router';
