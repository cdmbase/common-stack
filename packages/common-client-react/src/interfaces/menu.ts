

export interface IMenuData {
    [key: string]: {
        name: any;
        position?: IMenuPosition;
        icon?: any;
        authority?: any;
        exact?: boolean;
        strict?: boolean;
        [key: string]: any;
    };
}

export interface IMenuItem {
    path?: string;
    name: any;
    icon?: any;
    position: IMenuPosition;
    authority?: any;
    loading?: any;
    exact?: boolean;
    strict?: boolean;
    children?: IMenuItem[];
    priority?: any
}

export enum IMenuPosition {
    LOGO= 'LOGO',
    UPPER = 'UPPER',
    MIDDLE = 'MIDDLE',
    LOWER = 'LOWER',
    BOTTOM = 'BOTTOM',
}
