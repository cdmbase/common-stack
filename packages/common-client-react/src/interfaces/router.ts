import { LoadingComponentProps } from 'react-loadable';
import { any } from 'prop-types';
// import { IRouteData as IOrigROuteData } from '@common-stack/client-core';
// export { IMappedData } from '@common-stack/client-core';

export interface ICoreRouteData<T> {
    [key: string]: {
        loading?: T;
        component: any;
        exact?: boolean;
        strict?: boolean;
    };
}

export interface IMappedData {
    path?: string;
    component?: any;
    route?: string;
    exact?: boolean;
    strict?: boolean;
    [key: string]: any;
}

export interface IRouteData extends ICoreRouteData<React.ComponentType<LoadingComponentProps> | (() => null)> {
}
