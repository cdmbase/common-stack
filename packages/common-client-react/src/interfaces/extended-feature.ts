import * as React from 'react';
import {IFeature} from '@common-stack/client-core';
import {IMappedData, IRouteData} from './router';
import {ComponentElement} from 'react';

export interface IPlugin {
    name: string;
    icon?: string | ComponentElement<any, any> | Function;
    render: React.FC;
}

export type IReactFeature = IFeature & {
    routeConfig: IRouteData,
    menuConfig: IMappedData,
    // Related to React Slot Fill
    componentFillPlugins: IPlugin[];
    getComponentFillPlugins(): IPlugin[];
};
