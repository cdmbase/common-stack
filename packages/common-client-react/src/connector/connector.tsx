import * as React from 'react';
import {Route, Switch} from 'react-router-dom';
import {AbstractFeature} from '@common-stack/client-core';
import * as Logger from 'bunyan';
import {IPlugin, IReactFeature} from '../interfaces';
import {getMenus, getRoutes} from '../utils';
import {castArray, map, union, without} from 'lodash';
import {registerPlugin, getPlugins, getPlugin} from '../plugin-area/plugin-api';

const combine = (features, extractor) => without(union(...map(features,
        res => castArray(extractor(res)))), undefined);

export class Feature extends AbstractFeature implements IReactFeature {
    private logger;
    public componentFillPlugins;
    constructor(
        feature?: IReactFeature,
        // tslint:disable:trailing-comma
        ...features: IReactFeature[]
    ) {
        super(feature, ...features);
        this.logger = Logger.createLogger({
            name: Feature.name
        });
        this.componentFillPlugins = this.registerComponentFillPlugins(combine(arguments,
            (arg: IReactFeature) => arg.componentFillPlugins));
    }

    /* tslint:disable:jsx-no-lambda */
    private renderRoutes = (routes, solidRoutes, extraProps = {}, switchProps = {}) =>
        routes ? (
            <Switch {...switchProps}>
                {[
                    ...solidRoutes, ...routes.map((route, i) => (
                        <Route
                            key={route.key || i}
                            path={route.path}
                            exact={route.exact}
                            strict={route.strict}
                            render={props => (
                                <route.component {...props} {...extraProps} route={route}/>
                            )}
                        />
                    )),
                ]}
            </Switch>
        ) : null

    /**
     * Get the routes
     */
    public getRoutes(withRoot?: boolean, rootComponent?: any) {
        const configuredRoutes = this.getConfiguredRoutes();
        const solidRoutes = this.route.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {key: component.props.path}));
        return this.renderRoutes(configuredRoutes, solidRoutes);
    }

    /**
     * Get menus
     */
    public getMenus(withRoot?: boolean, rootComponent?: any) {
        return this.getConfiguredMenus();
    }

    /**
     * get configured routes.
     * Note: It overwrites the any duplicate key with latest loaded key.
     * TODO: Find a way to warn when there are duplicate keys.
     */
    public getConfiguredRoutes(searchPath = '/') {
        const routes = Object.assign({}, ...this.routeConfig);
        return getRoutes(searchPath, {...routes});
    }


    /**
     * get configured menus.
     * Note: It overwrites the any duplicate key with latest loaded key.
     * TODO: Find a way to warn when there are duplicate keys.
     */
    public getConfiguredMenus(searchPath = '/') {
        const routes = Object.assign({}, ...this.menuConfig);
        return getMenus(searchPath, {...routes});
    }

    get navItems() {
        return this.navItem.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    get navItemsRight() {
        return this.navItemsRight.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItems.length,
            }),
        );
    }

    public getWrappedRoot(root: React.ReactNode, req?: any): React.ReactNode {
        let nestedRoot = root;
        for (const componentFactory of this.rootComponentFactory) {
            nestedRoot = React.cloneElement(componentFactory(req), {}, nestedRoot);
        }
        return nestedRoot;
    }

    public getDataRoot(root) {
        let nestedRoot = root;
        for (const component of this.dataRootComponent) {
            nestedRoot = React.createElement(component, {}, nestedRoot);
        }
        return nestedRoot;
    }

    public registerLanguages(monaco) {
        for (const func of this.languagesFuncs) {
            func(monaco);
        }
    }

    private registerComponentFillPlugins(plugins): void {
        plugins.forEach((i) => {
            const {name} = i;
            const isFound = !!getPlugin(name);
            if (!isFound) {
                registerPlugin(name, i);
            }
        });
        return plugins;
    }

    public getComponentFillPlugins() {
        return getPlugins();
    }
}
