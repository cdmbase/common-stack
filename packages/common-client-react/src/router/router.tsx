import * as React from 'react';
import { Switch } from 'react-router-dom';
import { Feature } from '../connector';
import {IReactFeature} from '../interfaces';

const routerFactory = (routes) => <Switch>{routes}</Switch>;

export const FeatureWithRouterFactory = new Feature({
    routerFactory,
} as IReactFeature);
