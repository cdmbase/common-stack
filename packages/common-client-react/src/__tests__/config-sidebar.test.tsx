import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as Loadable from 'react-loadable';
import { Route } from 'react-router-dom';
import { getRoutes } from '../utils';
import { Feature } from '../connector';
import { FeatureWithRouterFactory } from '../router';
import { IMenuData } from '../interfaces';
import { isUrl, getMenus } from '../utils';
import 'jest';





const genMenuData: (namespace?: string) => IMenuData = (namespace = '') => ({
    [namespace + '/a']: { name: 'a', icon: 'a', authority: 'guest' },
    [namespace + '/a/1']: { name: '1', icon: '1' },
    [namespace + '/a/m']: { name: 'm', icon: 'm' },
    [namespace + '/b/b1']: { name: 'b1', icon: 'b1' },
    ['/b/b2']: { name: 'b2', icon: 'b2' },
    ['/a/m/a1']: { name: 'a1', icon: 'a1', category: 'middle' },
    [namespace + '/a/m/a2']: { name: 'a2', icon: 'a2', category: 'top' },
});

describe('getMenu utility with basic paths', () => {


    function formatter(data, parentAuthority = undefined) {
        return data.map(item => {
            let { path } = item;
            const result = {
                ...item,
                path,
                authority: item.authority || parentAuthority,
            };
            if (item.children) {
                result.children = formatter(item.children, item.authority || parentAuthority);
            }
            return result;
        });
    }

    function getFlatMenuData(menus) {
        let keys = {};
        menus.forEach(item => {
            if (item.children) {
                keys[item.path] = { ...item };
                keys = { ...keys, ...getFlatMenuData(item.children) };
            } else {
                keys[item.path] = { ...item };
            }
        });
        return keys;
    }

    test('with /a', () => {
        const data = getMenus('/', genMenuData());
        const formattedData = formatter(data);
        expect(data).toMatchSnapshot();
        expect(formattedData).toMatchSnapshot();

        expect(getFlatMenuData(formattedData)).toMatchSnapshot();
    });


    test('with given menu', () => {
        const genDashboardMenu = {
            ['/dashboard']: {},
            ['/dashboard/name']: {},
            ['/userinfo']: {},
            ['/userinfo/:id']: {},
            ['/userinfo/:id/info']: {},
        };
        const menu = [
            {
                path: '/dashboard',
                children: [
                    {
                        path: '/dashboard/name',
                    },
                ],
            },
            {
                path: '/userinfo',
                children: [
                    {
                        path: '/userinfo/:id',
                        children: [
                            {
                                path: '/userinfo/:id/info',
                            },
                        ],
                    },
                ],
            },
        ];

        const data = getMenus('/', genDashboardMenu as any);

        expect(data).toMatchObject(menu);

    });
});


describe('getMenu utility with connector', () => {
    it('with connector', () => {
        const connector = new Feature({ menuConfig: [genMenuData()]  });
        const connector2 = new Feature({ menuConfig: [genMenuData('/test')]  });

        const feature = new Feature(connector, connector2);
        const finalMenu = feature.getMenus();

        expect(finalMenu).toMatchSnapshot();
    });

});

