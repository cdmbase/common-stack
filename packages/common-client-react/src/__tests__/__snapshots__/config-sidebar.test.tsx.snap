// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMenu utility with basic paths with /a 1`] = `
Array [
  Object {
    "authority": "guest",
    "children": Array [
      Object {
        "exact": true,
        "icon": "1",
        "name": "1",
        "path": "/a/1",
        "position": "MIDDLE",
      },
      Object {
        "children": Array [
          Object {
            "category": "middle",
            "exact": true,
            "icon": "a1",
            "name": "a1",
            "path": "/a/m/a1",
            "position": "MIDDLE",
          },
          Object {
            "category": "top",
            "exact": true,
            "icon": "a2",
            "name": "a2",
            "path": "/a/m/a2",
            "position": "MIDDLE",
          },
        ],
        "exact": false,
        "icon": "m",
        "name": "m",
        "path": "/a/m",
        "position": "MIDDLE",
      },
    ],
    "exact": false,
    "icon": "a",
    "name": "a",
    "path": "/a",
    "position": "MIDDLE",
  },
  Object {
    "exact": true,
    "icon": "b1",
    "name": "b1",
    "path": "/b/b1",
    "position": "MIDDLE",
  },
  Object {
    "exact": true,
    "icon": "b2",
    "name": "b2",
    "path": "/b/b2",
    "position": "MIDDLE",
  },
]
`;

exports[`getMenu utility with basic paths with /a 2`] = `
Array [
  Object {
    "authority": "guest",
    "children": Array [
      Object {
        "authority": "guest",
        "exact": true,
        "icon": "1",
        "name": "1",
        "path": "/a/1",
        "position": "MIDDLE",
      },
      Object {
        "authority": "guest",
        "children": Array [
          Object {
            "authority": "guest",
            "category": "middle",
            "exact": true,
            "icon": "a1",
            "name": "a1",
            "path": "/a/m/a1",
            "position": "MIDDLE",
          },
          Object {
            "authority": "guest",
            "category": "top",
            "exact": true,
            "icon": "a2",
            "name": "a2",
            "path": "/a/m/a2",
            "position": "MIDDLE",
          },
        ],
        "exact": false,
        "icon": "m",
        "name": "m",
        "path": "/a/m",
        "position": "MIDDLE",
      },
    ],
    "exact": false,
    "icon": "a",
    "name": "a",
    "path": "/a",
    "position": "MIDDLE",
  },
  Object {
    "authority": undefined,
    "exact": true,
    "icon": "b1",
    "name": "b1",
    "path": "/b/b1",
    "position": "MIDDLE",
  },
  Object {
    "authority": undefined,
    "exact": true,
    "icon": "b2",
    "name": "b2",
    "path": "/b/b2",
    "position": "MIDDLE",
  },
]
`;

exports[`getMenu utility with basic paths with /a 3`] = `
Object {
  "/a": Object {
    "authority": "guest",
    "children": Array [
      Object {
        "authority": "guest",
        "exact": true,
        "icon": "1",
        "name": "1",
        "path": "/a/1",
        "position": "MIDDLE",
      },
      Object {
        "authority": "guest",
        "children": Array [
          Object {
            "authority": "guest",
            "category": "middle",
            "exact": true,
            "icon": "a1",
            "name": "a1",
            "path": "/a/m/a1",
            "position": "MIDDLE",
          },
          Object {
            "authority": "guest",
            "category": "top",
            "exact": true,
            "icon": "a2",
            "name": "a2",
            "path": "/a/m/a2",
            "position": "MIDDLE",
          },
        ],
        "exact": false,
        "icon": "m",
        "name": "m",
        "path": "/a/m",
        "position": "MIDDLE",
      },
    ],
    "exact": false,
    "icon": "a",
    "name": "a",
    "path": "/a",
    "position": "MIDDLE",
  },
  "/a/1": Object {
    "authority": "guest",
    "exact": true,
    "icon": "1",
    "name": "1",
    "path": "/a/1",
    "position": "MIDDLE",
  },
  "/a/m": Object {
    "authority": "guest",
    "children": Array [
      Object {
        "authority": "guest",
        "category": "middle",
        "exact": true,
        "icon": "a1",
        "name": "a1",
        "path": "/a/m/a1",
        "position": "MIDDLE",
      },
      Object {
        "authority": "guest",
        "category": "top",
        "exact": true,
        "icon": "a2",
        "name": "a2",
        "path": "/a/m/a2",
        "position": "MIDDLE",
      },
    ],
    "exact": false,
    "icon": "m",
    "name": "m",
    "path": "/a/m",
    "position": "MIDDLE",
  },
  "/a/m/a1": Object {
    "authority": "guest",
    "category": "middle",
    "exact": true,
    "icon": "a1",
    "name": "a1",
    "path": "/a/m/a1",
    "position": "MIDDLE",
  },
  "/a/m/a2": Object {
    "authority": "guest",
    "category": "top",
    "exact": true,
    "icon": "a2",
    "name": "a2",
    "path": "/a/m/a2",
    "position": "MIDDLE",
  },
  "/b/b1": Object {
    "authority": undefined,
    "exact": true,
    "icon": "b1",
    "name": "b1",
    "path": "/b/b1",
    "position": "MIDDLE",
  },
  "/b/b2": Object {
    "authority": undefined,
    "exact": true,
    "icon": "b2",
    "name": "b2",
    "path": "/b/b2",
    "position": "MIDDLE",
  },
}
`;

exports[`getMenu utility with connector with connector 1`] = `
Array [
  Object {
    "children": undefined,
    "component": undefined,
    "exact": true,
    "icon": "b1",
    "key": undefined,
    "name": "b1",
    "path": "/test/b/b1",
    "position": "MIDDLE",
    "priority": undefined,
    "tab": undefined,
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": undefined,
            "component": undefined,
            "exact": true,
            "icon": "a2",
            "key": undefined,
            "name": "a2",
            "path": "/test/a/m/a2",
            "position": "MIDDLE",
            "priority": undefined,
            "tab": undefined,
          },
        ],
        "component": undefined,
        "exact": false,
        "icon": "m",
        "key": undefined,
        "name": "m",
        "path": "/test/a/m",
        "position": "MIDDLE",
        "priority": undefined,
        "tab": undefined,
      },
      Object {
        "children": undefined,
        "component": undefined,
        "exact": true,
        "icon": "1",
        "key": undefined,
        "name": "1",
        "path": "/test/a/1",
        "position": "MIDDLE",
        "priority": undefined,
        "tab": undefined,
      },
    ],
    "component": undefined,
    "exact": false,
    "icon": "a",
    "key": undefined,
    "name": "a",
    "path": "/test/a",
    "position": "MIDDLE",
    "priority": undefined,
    "tab": undefined,
  },
  Object {
    "children": undefined,
    "component": undefined,
    "exact": true,
    "icon": "b2",
    "key": undefined,
    "name": "b2",
    "path": "/b/b2",
    "position": "MIDDLE",
    "priority": undefined,
    "tab": undefined,
  },
  Object {
    "children": undefined,
    "component": undefined,
    "exact": true,
    "icon": "b1",
    "key": undefined,
    "name": "b1",
    "path": "/b/b1",
    "position": "MIDDLE",
    "priority": undefined,
    "tab": undefined,
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": undefined,
            "component": undefined,
            "exact": true,
            "icon": "a2",
            "key": undefined,
            "name": "a2",
            "path": "/a/m/a2",
            "position": "MIDDLE",
            "priority": undefined,
            "tab": undefined,
          },
          Object {
            "children": undefined,
            "component": undefined,
            "exact": true,
            "icon": "a1",
            "key": undefined,
            "name": "a1",
            "path": "/a/m/a1",
            "position": "MIDDLE",
            "priority": undefined,
            "tab": undefined,
          },
        ],
        "component": undefined,
        "exact": false,
        "icon": "m",
        "key": undefined,
        "name": "m",
        "path": "/a/m",
        "position": "MIDDLE",
        "priority": undefined,
        "tab": undefined,
      },
      Object {
        "children": undefined,
        "component": undefined,
        "exact": true,
        "icon": "1",
        "key": undefined,
        "name": "1",
        "path": "/a/1",
        "position": "MIDDLE",
        "priority": undefined,
        "tab": undefined,
      },
    ],
    "component": undefined,
    "exact": false,
    "icon": "a",
    "key": undefined,
    "name": "a",
    "path": "/a",
    "position": "MIDDLE",
    "priority": undefined,
    "tab": undefined,
  },
]
`;
